packages/core/src/types.ts

```diff
@@ -75,6 +81,365 @@ export interface PositionWithOffset extends Position {
+
+	/**
+	 * The messages that the rule can report.
+	 */
+	messages: Record<string, string>;
+
+	/**
+	 * The visitor keys for the rule.
+	 */
+	visitorKeys?: Record<string, string[]>;
+
+	/**
+	 * The deprecated rules for the rule.
+	 */
+	deprecated?: boolean | undefined;
+
+	/**
+	 * When a rule is deprecated, indicates the rule ID(s) that should be used instead.
+	 */
+	replacedBy?: string[] | undefined;
+
+	/**
+	 * Indicates if the rule is fixable, and if so, what type of fix it provides.
+	 */
+	fixable?: RuleFixType | undefined;
+
+	/**
+	 * Indicates if the rule may provide suggestions.
+	 */
+	hasSuggestions?: boolean | undefined;
+}
+
+/**
+ * Represents the context object that is passed to a rule. This object contains
+ * information about the current state of the linting process and is the rule's
+ * view into the outside world.
+ */
+export interface RuleContext {
+	/**
+	 * The current working directory for the session.
+	 */
+	cwd: string;
+
+	/**
+	 * Returns the current working directory for the session.
+	 * @deprecated Use `cwd` instead.
+	 */
+	getCwd(): string;
+
+	/**
+	 * The filename of the file being linted.
+	 */
+	filename: string;
+
+	/**
+	 * Returns the filename of the file being linted.
+	 * @deprecated Use `filename` instead.
+	 */
+	getFilename(): string;
+
+	/**
+	 * The physical filename of the file being linted.
+	 */
+	physicalFilename: string;
+
+	/**
+	 * Returns the physical filename of the file being linted.
+	 * @deprecated Use `physicalFilename` instead.
+	 */
+	getPhysicalFilename(): string;
+
+	/**
+	 * The source code object that the rule is running on.
+	 */
+	sourceCode: SourceCode;
+
+	/**
+	 * Returns the source code object that the rule is running on.
+	 * @deprecated Use `sourceCode` instead.
+	 */
+	getSourceCode(): SourceCode;
+
+	/**
+	 * Shared settings for the configuration.
+	 */
+	settings: Record<string, unknown>;
+
+	/**
+	 * Parser-specific options for the configuration.
+	 * @deprecated Use `languageOptions.parserOptions` instead.
+	 */
+	parserOptions: Record<string, unknown>;
+
+	/**
+	 * The language options for the configuration.
+	 */
+	languageOptions: LanguageOptions;
+
+	/**
+	 * The CommonJS path to the parser used while parsing this file.
+	 * @deprecated No longer used.
+	 */
+	parserPath: string;
+
+	/**
+	 * The rule ID.
+	 */
+	id: string;
+
+	/**
+	 * The rule's configured options.
+	 */
+	options: unknown[];
+
+	/**
+	 * The report function that the rule should use to report problems.
+	 * @param violation The violation to report.
+	 */
+	report(violation: ViolationReport): void;
+}
+
+// #region Rule Fixing
+
+/**
+ * Manager of text edits for a rule fix.
+ */
+export interface RuleTextEditor {
+	/**
+	 * Inserts text after the specified node or token.
+	 * @param nodeOrToken The node or token to insert after.
+	 * @param text The edit to insert after the node or token.
+	 */
+	insertTextAfter(nodeOrToken: object, text: string): RuleTextEdit;
```

Comment:
The PRs in this repo are probably the heaviest users of `: object` I've seen this year 😄. I'm getting the impression types using them are split across two use cases:

* "Base" (_generalized_) types meant to be the constraints on type parameters, such as the current `RuleVisitor`
* _Generic_ types where there's a clear injectable, shared type

For "Base" (_generalized)_ types, it makes sense to use types like `object` and `unknown`. It's also common to name them something like `Any*` or `*Base` to make it clear that they're base types, not the ones end users directly use. Some of the types here are named like that, but some aren't - is that intentional?

For _generic_ types, it'd be more typical to use a type parameter. That way consumers of the type can substitute in their intended types, instead of having to do some `extends` strategy. Which would be very inconvenient for types like this with a lot of `: object`s!

Is there anything blocking making this type generic?

```suggestion
export interface RuleTextEditor<NodeOrToken> {
	/**
	 * Inserts text after the specified node or token.
	 * @param nodeOrToken The node or token to insert after.
	 * @param text The edit to insert after the node or token.
	 */
	insertTextAfter(nodeOrToken: NodeOrToken, text: string): RuleTextEdit;
```

Note that the same logic could be extended to any place that includes `: object` or `: unknown`.

* `ViolationLocation`'s `node`: which would make `ViolationReport` generic, which would make `RuleContext` generic.
* `RuleContext`'s `options` 
* `RuleVisitor`'s `node` and `parent` - though that type looks like it might be more of a "Base" _generalized_ type?

Reply:
In this case, there's only ever one instance of `RuleTextEditor`, which exists in the core and needs to work on everything...would that still be a case for a generic? 🤔 

(Still trying to wrap my brain around this.)

Reply:
Type parameters are for when you want to substitute out different parts of a type for different uses. If there's only ever one instance of something, then you likely don't need type parameters - but then the type for `nodeOrToken` should be known as something more specific than `object`, right?

If the type for `nodeOrToken` is always going to be some known thing, then a different suggestion could be:

```suggestion
// or some existing type, or some union thereof...
export interface EditableNodeOrToken {
  /* ... */
}

export interface RuleTextEditor {
	/**
	 * Inserts text after the specified node or token.
	 * @param nodeOrToken The node or token to insert after.
	 * @param text The edit to insert after the node or token.
	 */
	insertTextAfter(nodeOrToken: EditableNodeOrToken, text: string): RuleTextEdit;
```

Reply:
To fully flesh out what we have here: this same `RuleTextEditor` is used inside the core for applying fixes to any language that ESLint is linting (via a `Language` object). So for JavaScript, it works on ESTree(ish) nodes, while for Markdown it works on Unist nodes. While these two types of nodes only shared a `type`, there is no guarantee that any particular language will have that.

So, a rule is passing some representation of a node (or token) into these methods on `RuleTextEditor`. ESLint itself doesn't really need to know what those are, because we just pass them back into `Language#getRange()` to get the range information for that node or token. 

🤔 I suppose the audience for types here would be the rule developer, who may not be able to tell what to pass in. In that case, I could see it being helpful to specify the type as a parameter...but that would then bubble up into `RuleContext` and then up to `RuleDefinition`. Maybe that's not a bad thing in TypeScript land?

Reply:
👍 if the `RuleContext` cares about the type of nodes, and `RuleDefinition` therefore does as well, then it makes sense that `RuleDefinition` would have a type parameter for the kind of nodes.

In other words: if some consumer of `RuleContext`/`RuleDefinition` wanted to write a function that takes one of those in...

```js
// js version
function workOnContextNode(context: RuleContext) {
 context.sourceCode.ast.exampleProperty; // (or some equivalent)
}
```

...then TS users will want to be able to have that be well-typed without an `as`.


```js
interface MyLanguageRootAST {
  exampleProperty: true;
}

function workOnContextNode(context: RuleContext<MyLanguage>) {
 context.sourceCode.ast.exampleProperty;
}
```

Reply:
I think so. Going to dig into this now a bit more.